### ISSUE TITLE
autoretry_for=(Exception,) creates extra retries when Retry exception raised 

### ISSUE THREAD

**Issue by lsemel** [celery/celery#4898]
I added `autoretry_for=(Exception,)` as described in the docs, to a task that itself would sometimes issue a regular `raise self.retry()` under certain conditions.   This actually causes multiple retries to be raised, creating an explosion of tasks.

Since `autoretry_for` requires a tuple of exceptions, I don't see an easy way to exclude the `Retry` exception itself while allowing for all other exceptions.  

## Steps to reproduce

```
from celery import app
from django.test import TestCase
from django.test import override_settings


@app.task(bind=True, autoretry_for=(Exception,), max_retries=10)
def manual_retry_task(self):
    if not hasattr(self, 'count'):
        self.count = 0
    self.count += 1
    raise self.retry()

@override_settings(CELERY_TASK_ALWAYS_EAGER=True)
class TestTask(TestCase):
    def test_retry(self):
        manual_retry_task.delay()
        print manual_retry_task.count
```

## Expected behavior

I'd have expected autoretry_for to ignore any `Retry` exceptions raised manually within the body of the task, and for the task to execute 11 times

## Actual behavior

Task executes 2047 times.  The same thing happens when run with a real worker outside of eager mode.

**Comment by lsemel** [comment]
Within `_task_from_fun `in `celery/app/base.py`, adding an exception handler that catches Retry and re-raises it solves this for a real worker, but in eager mode the underlying exception is a RuntimeError due to a .get() call from within the body of the task.

**Comment by auvipy** [comment]
so whats the resolution? should this be closed as a duplicate of #4661?

**Comment by lsemel** [comment]
No, it's different than #4661 - that relates specifically to retries in eager mode, this situation happens with any retries within a task that uses the autoretry_for functionality

**Comment by auvipy** [comment]
thanks for clarifying.

**Comment by auvipy** [comment]
could you send a PR with test?

**Comment by xirdneh** [comment]
Isn't the issue here that the task is trying to retry for any exception by using `auto_retry=(Exception,)` isntesd of using a smaller subset of exception classes?
Since `Retry` is itself a subclass of `Exception` then it will retry.

If this is correct then I would suggest not to modify the code to explicitly ignore `Retry` since this might create some issues in the future.

**Comment by cockcrow** [comment]
I got a similar issue when using `autoretry_for=(Exception,)` and `Reject` with `requeue=False` together. The original idea was to auto retry for most cases, and reject for  some known issues.
Maybe opposite to the design ideas.

**Comment by mchataigner** [comment]
@auvipy this may have been partially fixed with https://github.com/celery/celery/pull/6138. I say partially because in my PR I only take care of Retry and not MaxRetriesExceededError.

**Comment by auvipy** [comment]
thanks for triaging! I hope you will handle the remaining edge case

**Comment by sjaq** [comment]
Is this still an issue? 

If it is I would like to work on it, but I'm totally unfamiliar with the celery codebase. Would the changes in #6138 be a good place to look if I want to fix this behavior?

IMHO it's a fairly common approach to retry on "most" errors and instead write task code that catches the known issues instead of the other way around, so I guess I wouldn't be alone in this?

**Comment by auvipy** [comment]
I am open to review any draft PR
